src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs

```diff
@@ -205,7 +205,28 @@ private BoundExpression ConvertCaseExpression(TypeSymbol switchGoverningType, CS
 
             if (switchGoverningType.IsNullableType() && convertedCaseExpression.Kind == BoundKind.Conversion)
             {
-                constantValueOpt = ((BoundConversion)convertedCaseExpression).Operand.ConstantValue;
+                var operand = ((BoundConversion)convertedCaseExpression).Operand;
+
+                if (operand.ConstantValue.IsIntegral)
+                {
+                    // We have to make sure we convert numerics correctly
+                    //public static bool F(long? x)
+                    //{
+                    //    switch (x)
+                    //    {
+                    //        // '1' has to be converted from Int32 to Int64 here
+                    //        case 1:
+                    //            return true;
+                    //        default:
+                    //            return false;
+                    //    }
+                    //}
+                    constantValueOpt = FoldConstantConversion(convertedCaseExpression.Syntax, operand, Conversion.ImplicitNumeric, switchGoverningType.GetNullableUnderlyingType(), diagnostics);
```

Comment:
Rather than guessing what the conversion should be (Conversion.ImplicitNumeric), I would suggest to bind conversion from the operand to the new target type as we normally would and getting constant value from the result. It also feels that that should be done unconditionally, regardless whether ```operand.ConstantValue.IsIntegral```.